pallets/confidential-asset/src/lib.rs

```diff
@@ -36,19 +51,91 @@ pub trait Trait:
+/// not automatically implemented.
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq, VecU8StrongTyped, Default)]
+pub struct EncryptedBalanceWrapper(pub Vec<u8>);
+
+/// A mercat account consists of the public key that is used for encryption purposes and the
+/// encrypted asset id. The encrypted asset id also acts as the unique identifier of this
+/// struct.
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq, Default)]
+pub struct MercatAccount {
+    pub encrypted_asset_id: EncryptedAssetIdWrapper,
+    pub encryption_pub_key: EncryptionPubKey,
+}
+
 decl_storage! {
-    trait Store for Module<T: Trait> as Asset {
+    trait Store for Module<T: Trait> as ConfidentialAsset {
+
+        /// Contains the mercat accounts for an identity.
+        pub MercatAccounts get(fn mercat_account): double_map hasher(blake2_128_concat) IdentityId, hasher(blake2_128_concat) EncryptedAssetIdWrapper  => MercatAccount;
+
+        /// Contains the encrypted balance of a mercat account.
+        pub MercatAccountBalance get(fn mercat_account_balance):  double_map hasher(blake2_128_concat) IdentityId, hasher(blake2_128_concat) EncryptedAssetIdWrapper =>EncryptedAssetIdWrapper;
+
+        /// Contains the list of all valid ticker names.
+        /// The process around creating and storing this data will likely change as a result of CRYP-153.
+        pub ValidAssetIds get(fn valid_asset_ids): Vec<AssetId>;
     }
 }
 
 // Public interface for this runtime module.
 decl_module! {
     pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+
+        /// TODO: CRYP-153 will most likely change it into a private function that is called
+        /// when a new ticker name is registered.
+        #[weight = 1_000_000_000]
+        pub fn store_valid_asset_ids(
+            origin,
+            valid_asset_ids: Vec<AssetId>,
+        ) -> DispatchResult
+        {
+            ensure_signed(origin)?;
+
+            <ValidAssetIds>::put(valid_asset_ids);
+            Ok(())
+        }
+
+        /// It receives the transaction data for creating a mercat account and verifies the proofs.
+        /// If all proofs pass, it stores the mercat account for the origin identity.
+        #[weight = 1_000_000_000]
+        pub fn validate_mercat_account(origin,
+            tx: PubAccountTx,
+        ) -> DispatchResult
+        {
+            let owner_acc = ensure_signed(origin)?;
+            let owner_id = Context::current_identity_or::<Identity<T>>(&owner_acc)?;
+
+            let valid_asset_ids = Self::valid_asset_ids();
+            let valid_asset_ids = convert_asset_ids(valid_asset_ids);
+            AccountValidator{}.verify(&tx, &valid_asset_ids).map_err(|_| Error::<T>::InvalidAccountCreationProof)?;
+            let wrapped_enc_asset_id = EncryptedAssetIdWrapper::from(tx.pub_account.enc_asset_id.encode());
+            <MercatAccounts>::insert(&owner_id, &wrapped_enc_asset_id.clone(), MercatAccount {
```

Comment:
I would suggest adding `new()` function to create the instance of the `MercatAccount`

i.e - 

```
impl MercatAccount {
  pub fn new(encrypted_asset_id, encryption_pub_key) -> Self {
   ....
}
}

Reply:
Why? I thought that the Rust way was to use new only if the content of the struct are private or if we need to do extra processing in the new function. At the moment it is only passing the two values to the two public members of the struct.

Reply:
I am recommending this because if we use the same functionality in other places as well then it will be nice to have this functionality as it is more readable if you think this will not be used anywhere then we can keep the same.

Reply:
Thanks @satyamakgec, I am pretty sure this will be the only place. But, I'll definitely change it if we end up using it in other places :)